### DISCUSSION TITLE
Using different datacontext inside one query

### DISCUSSION THREAD

**Discussion by ademchenko** [linq2db/linq2db#4711] [Q&A] [answered]
Let's say for some reason in my interfaces I want to work with IQueryables without exposing internal details of how they were built, so I would have two methods used DataContexts internally:
```
IQueryable<User> GetUsers() => new DataContext().GetTable<User>();
IQueryable<SalaryOfUser> GetSalaryOfUser() => new DataContext().GetTable<SalaryOfUser>();
```

And now I want to join the information about users and their salaries:
```
public IQueryable<UserSalaryInfo> GetUsersWithSalaries()
{
    var users = GetUsers();            // IQueryable<User>
    var salaries = GetSalaryOfUser();   // IQueryable<SalaryOfUser>

    var query = from user in users
                join salary in salaries
                on user.UserId equals salary.UserId
                select new UserSalaryInfo
                {
                    UserId = user.UserId,
                    UserName = user.UserName,
                    Salary = salary.Amount
                };

    return query;
}
```
The question is - when I materialize those IQueryable\<UserSalaryInfo\> with, let's say, ToListAsync() from linq2db would that be a problem that users and salaries are built over the different datacontexts?

**Comment by jods4** [comment, marked as answer]
Yes, it would as they represent two different DB connections and you can't build a single query across two connections.

You have two options here:
1. Share the datacontext. Opening new datacontexts as you do in this example is a bad practice. Your datacontext is basically your DB connection, you only want one.
```cs
private DataContext context;

IQueryable<User> GetUsers() => context.GetTable<User>();
IQueryable<SalaryOfUser> GetSalaryOfUser() => context.GetTable<SalaryOfUser>();
```

2. If you insist on having 2 connections, then you must join the results in memory. You can force this by doing `AsEnumerable()` or `ToList()` on your queryables. You could also do this on only one side (e.g. users), then the second query would be some kind of `userId IN (1, 2, 4)` with the results of the first query. 
Both variants are bad though: first one might load a lot of useless data in memory because filters are applied too late; second has limits on how many values you can pass in a query like that.
